src/translator/cache.cpp

```diff
@@ -0,0 +1,157 @@
+      cacheIdentifier, L4::HashTableConfig::Setting{static_cast<uint32_t>(numBuckets)}, cacheConfig_));
+}
+
+bool ThreadSafeL4Cache::fetch(const marian::Words &words, ProcessedRequestSentence &processedRequestSentence) {
+  auto &hashTable = context_[hashTableIndex_];
+
+  /// We take marian's default hash function, which is templated for hash-type. We treat marian::Word (which are indices
+  /// represented by size_t (unverified) and convert it into a uint64_t for hashing with more bits)
+  ///
+  /// Note that there can still be collisions, in which case the translations returned from cache can be wrong. It's
+  /// practically very unlikely (claims @XapaJIaMnu). To guarantee correctness, comparisons need to be made with the
+  /// actual key, which is cannot be done with this approach, as the cache is unaware of the original marian::Words key.
+  //
+  /// TODO(@jerinphilip): Empirical evaluation that this is okay.
+
+  KeyBytes keyBytes;
+  auto key = hashFn_(words);
+
+  // L4 requires uint8_t byte-stream, so we treat 64 bits as a uint8 array, with 8 members.
+  keyBytes.m_data = reinterpret_cast<const std::uint8_t *>(&key);
+  keyBytes.m_size = 8;
+
+  ValueBytes valBytes;
+  bool fetchSuccess = hashTable.Get(keyBytes, valBytes);
+  if (fetchSuccess) {
+    const char *data = reinterpret_cast<const char *>(valBytes.m_data);
+    size_t size = valBytes.m_size;
+    string_view bytesView(data, size);
+    processedRequestSentence = std::move(ProcessedRequestSentence::fromBytesView(bytesView));
+  }
+
+  return fetchSuccess;
+}
+
+void ThreadSafeL4Cache::insert(const marian::Words &words, const ProcessedRequestSentence &processedRequestSentence) {
+  auto context = service_.GetContext();
+  auto &hashTable = context[hashTableIndex_];
+
+  KeyBytes keyBytes;
+
+  auto key = hashFn_(words);
+
+  // L4 requires uint8_t byte-stream, so we treat 64 bits as a uint8 array, with 8 members.
+  keyBytes.m_data = reinterpret_cast<const std::uint8_t *>(&key);
+  keyBytes.m_size = 8;
+
+  ValueBytes valBytes;
+  string_view serialized = processedRequestSentence.toBytesView();
+
+  valBytes.m_data = reinterpret_cast<const std::uint8_t *>(serialized.data());
+  valBytes.m_size = sizeof(string_view::value_type) * serialized.size();
+
+  hashTable.Add(keyBytes, valBytes);
+}
+
+CacheStats ThreadSafeL4Cache::stats() const {
+  auto &perfData = context_[hashTableIndex_].GetPerfData();
+  CacheStats stats;
+  stats.hits = perfData.Get(L4::HashTablePerfCounter::CacheHitCount);
+  stats.misses = perfData.Get(L4::HashTablePerfCounter::CacheMissCount);
+  stats.activeRecords = perfData.Get(L4::HashTablePerfCounter::RecordsCount);
+  stats.evictedRecords = perfData.Get(L4::HashTablePerfCounter::EvictedRecordsCount);
+  stats.totalSize = perfData.Get(L4::HashTablePerfCounter::TotalIndexSize);
+  return stats;
+}
+
+#endif  // WASM_COMPATIBLE_SOURCE
+
+ThreadUnsafeLRUCache::ThreadUnsafeLRUCache(Ptr<Options> options)
```

Comment:
Just thoughts, thinking about real world usecases. We probably don't particularly care if the cache is LRU, because the principal usecases are a person editing something they've input, or, alternatively, loading additional page content. Good to benchmark the cost though.